src/realm/sync/instruction_applier.cpp

```diff
@@ -1,30 +1,58 @@
 #include <realm/sync/instruction_applier.hpp>
 #include <realm/sync/object.hpp>
 #include <realm/set.hpp>
+#include <realm/util/scope_exit.hpp>
 
 #include <realm/group.hpp>
 
 namespace realm::sync {
 
-namespace {
-REALM_NORETURN void throw_bad_transaction_log(std::string msg)
+REALM_NORETURN void InstructionApplier::bad_transaction_log(const std::string& msg) const
 {
-    throw BadChangesetError{std::move(msg)};
-}
-REALM_NORETURN void bad_transaction_log(const char* msg)
-{
-    throw_bad_transaction_log(msg);
+    if (m_last_object_key) {
+        // If the last_object_key is valid then we should have a changeset and a current table
+        REALM_ASSERT(m_log);
+        REALM_ASSERT(m_last_table_name);
+        std::stringstream ss;
+        util::Optional<InternString> field_name;
+        if (m_last_field_name) {
+            field_name = m_last_field_name;
+        }
+        const instr::Path* cur_path = m_current_path ? &(*m_current_path) : nullptr;
+        m_log->print_path(ss, m_last_table_name, *m_last_object_key, field_name, cur_path);
+        throw BadChangesetError{
+            util::format("%1 (instruction target: %2, version: %3, last_integrated_remote_version: %4, "
```

Comment:
The object id and path do not apply for errors relating to some niche cases such as: create object, add column -> bad changeset. But I think that is rare enough that this is overall a net positive as is and people reading the logs at this level of detail should be able to figure it out.